src/components/LayoutTree.tsx

```diff
@@ -361,175 +366,185 @@ const linearizeTree = (
-  bulletWidth: number
-  cursorUncleId: string | null
-  setSize: OnResize
-  cliffPaddingStyle: { paddingBottom: number }
-  dragInProgress: boolean
-  autofocusDepth: number
-}) => {
-  const [y, setY] = useState(_y)
-
-  useLayoutEffect(() => {
-    if (y !== _y) {
-      // When y changes React re-renders the component with the new value of y. It will result in a visual change in the DOM.
-      // Because this is a state-driven change, React applies the updated value to the DOM, which causes the browser to recognize that
-      // a CSS property has changed, thereby triggering the CSS transition.
-      // Without this additional render, updates get batched and subsequent CSS transitions may not work properly. For example, when moving a thought down, it would not animate.
-      setY(_y)
-    }
-  }, [y, _y])
-
-  // List Virtualization
-  // Do not render thoughts that are below the viewport.
-  // Exception: The cursor thought and its previous siblings may temporarily be out of the viewport, such as if when New Subthought is activated on a long context. In this case, the new thought will be created below the viewport and needs to be rendered in order for scrollCursorIntoView to be activated.
-  // Render virtualized thoughts with their estimated height so that document height is relatively stable.
-  // Perform this check here instead of in virtualThoughtsPositioned since it changes with the scroll position (though currently `sizes` will change as new thoughts are rendered, causing virtualThoughtsPositioned to re-render anyway).
-  if (belowCursor && !isCursor && y > viewportBottom + height) return null
-
-  const nextThought = isTableCol1 ? treeThoughtsPositioned[index + 1] : null
-  const previousThought = isTableCol1 ? treeThoughtsPositioned[index - 1] : null
+const TreeNode = forwardRef<
+  HTMLDivElement,
+  TreeThoughtPositioned & {
+    thoughtKey: string
+    index: number
+    viewportBottom: number
+    treeThoughtsPositioned: TreeThoughtPositioned[]
+    bulletWidth: number
+    cursorUncleId: string | null
+    setSize: OnResize
+    cliffPaddingStyle: { paddingBottom: number }
+    dragInProgress: boolean
+    autofocusDepth: number
+  }
+>(
+  (
+    {
+      belowCursor,
+      cliff,
+      depth,
+      env,
+      height,
+      indexChild,
+      indexDescendant,
+      isCursor,
+      isTableCol1,
+      isTableCol2,
+      thoughtKey,
+      leaf,
+      path,
+      prevChild,
+      showContexts,
+      isLastVisible,
+      simplePath,
+      singleLineHeightWithCliff,
+      style,
+      thoughtId,
+      width,
+      autofocus,
+      x,
+      y: _y,
+      index,
+      viewportBottom,
+      treeThoughtsPositioned,
+      bulletWidth,
+      cursorUncleId,
+      setSize,
+      cliffPaddingStyle,
+      dragInProgress,
+      autofocusDepth,
+    },
+    ref,
+  ) => {
+    const [y, setY] = useState(_y)
+
+    useLayoutEffect(() => {
+      if (y !== _y) {
```

Comment:
> If the new value you provide is identical to the current `state`, as determined by an [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) comparison, React will **skip re-rendering the component and its children**. This is an optimization. Although in some cases React may still need to call your component before skipping the children, it shouldn’t affect your code.

https://react.dev/reference/react/useState#setstate-caveats

I still think the `if (y !== _y)` check is unnecessary because React will not re-render if `y === _y` regardless. However, it doesn't hurt to be explicit with the intent. I am okay with this as-is.